x/exchange/keeper/market.go

```diff
@@ -0,0 +1,1303 @@
+		errs = append(errs, err)
+	}
+	if len(errs) > 0 {
+		return errors.Join(errs...)
+	}
+
+	k.emitEvent(ctx, exchange.NewEventMarketReqAttrUpdated(marketID, admin))
+	return nil
+}
+
+// getMarketAccountByAddr gets a market's account given its address.
+// This is for when you've already called exchange.GetMarketAddress(marketID) and need it for other things too.
+func (k Keeper) getMarketAccountByAddr(ctx sdk.Context, marketAddr sdk.AccAddress) *exchange.MarketAccount {
+	acc := k.accountKeeper.GetAccount(ctx, marketAddr)
+	if acc == nil {
+		return nil
+	}
+	marketAcc, ok := acc.(*exchange.MarketAccount)
+	if !ok {
+		return nil
+	}
+	return marketAcc
+}
+
+// GetMarketAccount gets a market's account from the account module.
+func (k Keeper) GetMarketAccount(ctx sdk.Context, marketID uint32) *exchange.MarketAccount {
+	marketAddr := exchange.GetMarketAddress(marketID)
+	return k.getMarketAccountByAddr(ctx, marketAddr)
+}
+
+// GetMarketDetails gets a market's details.
+func (k Keeper) GetMarketDetails(ctx sdk.Context, marketID uint32) *exchange.MarketDetails {
+	marketAcc := k.GetMarketAccount(ctx, marketID)
+	if marketAcc == nil {
+		return nil
+	}
+	return &marketAcc.MarketDetails
+}
+
+// UpdateMarketDetails updates a market's details. It returns an error if the market account
+// isn't found or if there aren't any changes provided.
+func (k Keeper) UpdateMarketDetails(ctx sdk.Context, marketID uint32, marketDetails exchange.MarketDetails, updatedBy sdk.AccAddress) error {
+	if err := marketDetails.Validate(); err != nil {
+		return err
+	}
+
+	marketAcc := k.GetMarketAccount(ctx, marketID)
+	if marketAcc == nil {
+		return fmt.Errorf("market %d account not found", marketID)
+	}
+
+	if marketAcc.MarketDetails.Equal(marketDetails) {
+		return errors.New("no changes")
+	}
+
+	marketAcc.MarketDetails = marketDetails
+	k.accountKeeper.SetAccount(ctx, marketAcc)
+	k.emitEvent(ctx, exchange.NewEventMarketDetailsUpdated(marketID, updatedBy))
+
+	return nil
+}
+
+// storeMarket writes all the market fields to the state store (except MarketDetails which are in the account).
+func storeMarket(store sdk.KVStore, market exchange.Market) {
+	marketID := market.MarketId
+	setMarketKnown(store, marketID)
+	setCreateAskFlatFees(store, marketID, market.FeeCreateAskFlat)
+	setCreateBidFlatFees(store, marketID, market.FeeCreateBidFlat)
+	setSellerSettlementFlatFees(store, marketID, market.FeeSellerSettlementFlat)
+	setSellerSettlementRatios(store, marketID, market.FeeSellerSettlementRatios)
+	setBuyerSettlementFlatFees(store, marketID, market.FeeBuyerSettlementFlat)
+	setBuyerSettlementRatios(store, marketID, market.FeeBuyerSettlementRatios)
+	setMarketActive(store, marketID, market.AcceptingOrders)
+	setUserSettlementAllowed(store, marketID, market.AllowUserSettlement)
+	setAccessGrants(store, marketID, market.AccessGrants)
+	setReqAttrsAsk(store, marketID, market.ReqAttrCreateAsk)
+	setReqAttrsBid(store, marketID, market.ReqAttrCreateBid)
```

Comment:
Can we store some of these together? Is this being done for optimization purposes? It just seems like a lot to have getters, setters, and iterators for.

Reply:
It's done for optimization purposes so that as we're processing things, we only have to look up the pieces of data that are needed.

Reply:
I reviewed some of the msg_server endpoints to follow the logic through.  Some of them are doing a lot OF individual reads from the store, i.e. `CreateAsk` and `CreateBid`.  I understand what @Taztingo means when asking if grouping these are better.  
Also, my other concern is maintainability, I didn't really understand the relationship of the different parts until I stepped through the keeper.  Grouping them would have made it a little easier to understand that `AskFlatFees`, `SellerSettlementFlatFees`,  `setReqAttrsAsk` are going to be used together.  Also, maybe it would reduce the amount of code/tests to maintain as well?

Reply:
Even though creating an ask will require lookup of required create-ask attributes, and create-ask flat fees, we only need to check one fee denom (the one provided). Yes, there are more lookups, but the amount of data looked up is significantlly less than if things were bundled together.

We have a problem with markers and how they're starting to have too much data in them. E.g. in order to check the permissions of one address, you have to read all of the marker's permissions for all addresses first. We've already had to address a similar problem in the SDK with the send-enabled list.

So yes, it is an extra burden too build, but it's a necessary burden. As for maintenance, it's pretty simple: Each aspect has its own getters/setters; when you need to get something, you get just that thing, when you need to update something, you update just that thing. There isn't any duplicated logic code either. So when a thing needs changing, you can change just that thing.

Since I haven't written the unit tests on most of the keeper stuff, I also haven't audited what's there vs what's actually being used. So there's a good chance I'll be able to clean up some of that stuff, but It still needs to be broken out like this.